pagination/pagination.go

```diff
@@ -0,0 +1,129 @@
+package pagination
+
+import (
+	"encoding/json"
+	"fmt"
+	"math"
+	"net/http"
+	"net/url"
+	"strconv"
+)
+
+const defaultPageNumber = 1
+const defaultItemsPerPage = 10
+
+// Pagination contains all of the JSON response fields as well as a url
+// that is later used in the json encoding to populate the hrefs dynamiclly
+// some strings are pointers so they can be a null JSON value if empty
+type Pagination struct {
+	PageNumber        int     `json:"page_number"`
+	PagesTotal        int     `json:"pages_total"`
+	ItemsPerPage      int     `json:"items_per_page"`
+	ItemsPerPageLimit int     `json:"items_per_page_limit"`
+	ItemsTotal        int     `json:"items_total"`
+	FirstHref         string  `json:"first_href"`
+	LastHref          string  `json:"last_href"`
+	NextHref          *string `json:"next_href"`
+	PrevHref          *string `json:"prev_href"`
+	request           *http.Request
+}
+
+// GetOffset returns the current offset (i.e. for database queries)
+// based on the current page and number of itmes
+func (p *Pagination) GetOffset() int {
+	if p.PageNumber == 1 {
+		return 0
+	}
+	return (p.PageNumber - 1) * p.ItemsPerPage
+}
+
+// SetItemsTotal will set the ItemsTotal value as well as the dynamic PagesTotal based on ItemsPerPage
+func (p *Pagination) SetItemsTotal(i int) {
+	p.ItemsTotal = i
+	p.PagesTotal = int(math.Ceil(float64(i) / float64(p.ItemsPerPage)))
+}
+
+// getURL is a helper used to build the HREFs based on the current URL
+// it really just overrides the currently set page query string with the required value
+func (p *Pagination) getURL(page int) (u *url.URL) {
+	u = p.request.URL
+
+	q := u.Query()
+	q.Set("page", strconv.Itoa(page))
+	q.Set("limit", strconv.Itoa(p.ItemsPerPage))
+	u.RawQuery = q.Encode()
+
+	// The URL does not always contain the required information, so if its not there
+	// build it from the request info
+	if "" == u.Host {
+		u.Host = p.request.Host
+	}
+	if "" == u.Scheme {
+		u.Scheme = "http"
+		if nil != p.request.TLS {
+			u.Scheme = "https"
+		}
+	}
+
+	return
+}
+
+// New returns a configured Pagination struct that can be used
+func New(PageNumber int, ItemsPerPage int, ItemsPerPageLimit int, r *http.Request) (p Pagination, err error) {
```

Comment:
I get what you are saying, but I don't think the Pagination should be responsible for fetching and parsing input parameters, it is outside of its purpose and provides more restrictions on the user (we ignore their input filtering/validation, we force the names of URL params). The implementation of Pagination is clearly laid out as "constructor params" required use this.

Having said that I realise I am already forcing the use of specific names through the pageURL() function, although this could be quite easily fixed if needed at a later date by providing default url param names and the ability to override them.